### PR TITLE
Display list of droplets before deleting

### PR DESCRIPTION
**Pull request by bnordland**
For issue #172, this will show the list of droplets being deleted if we
aren’t forcing so that the confirmation has more information for us.
Note: does not handle printing the list when doing a DeleteByTag.

We want to make sure that the list shown before the confirmation is actually the list that will be deleted which is why we retrieve the list once and use the same list before and after the confirmation. That way if the confirmation had been up and someone else deletes and recreates droplet "test" as a different one, since it is working off of IDs we don't delete one that was not shown.
If we wanted to do deleteByTag as well we would probably want a similar mechanism to do ListByTag, show the confirmation, and then loop through the same list after the confirmation doing the ds.Delete from the original list rather than ds.DeleteByTag directly.

### CODE REVIEW

**Comment by xmudrii** [comment]
LGTM, but you need LGTM from some maintainer. :D

cc @aybabtme @nanzhong

File: commands/droplets.go [hunk 1]
```diff
@@ -475,19 +475,39 @@ func RunDropletDelete(c *CmdConfig) error {
...
 	}
 
-	if force || AskForConfirm("delete droplet(s)") == nil {
-
-		fn := func(ids []int) error {
+	fn := func(ids []int) error {
+		if !force {
+		
+			// If we aren't forcing, print out the list
+			fmt.Printf("Following Droplet(s) will be deleted:\n")
```
Comment (xmudrii) [inline commands/droplets.go]: Maybe a tiny nitpick, but this could be changed with `Println` (in that case `\n` is not needed).

Comment (bnordland) [inline commands/droplets.go]: Good point! I'll make that change.

File: commands/droplets_test.go [hunk 2]
```diff
@@ -160,7 +160,9 @@ func TestDropletDelete(t *testing.T) {
 
 func TestDropletDeleteByTag(t *testing.T) {
 	withTestClient(t, func(config *CmdConfig, tm *tcMocks) {
-		tm.droplets.On("DeleteByTag", "my-tag").Return(nil)
```
Comment (xmudrii) [inline commands/droplets_test.go]: What is the reason behind this change? This function is made to test DeleteByTag and I think it should stay as it is, but if there is good reason behind it, it can be changed.

**Comment by xmudrii** [comment]
As #180 got accepted, does we need a rebase now?

File: commands/droplets_test.go (hunk 2 above)
Comment (bnordland) [inline commands/droplets_test.go]: Since DeleteByTag now actually under the covers does a ListByTag, and then a delete on the returned list items by id, this means that the test needs to have mocks for ListByTag, and Delete instead of calling the droplet service DeleteByTag.

The reason that I changed the actual implementation to do a ListByTag and then do the deletes separately instead of the obvious answer of:
ListByTag
Show confirmation
DeleteByTag

Is to prevent the following from happening:
1. The user does a DeleteByTag named "test" and is shown a list of three droplets to delete.
2. something else, either the user in another process or script or something tags another droplet under the same tag, so there are now four droplets tagged with the tag "test"
3. The user confirms the DeleteByTag request. But instead of the original three droplets they were shown four are deleted which could lead to inadvertent data loss.

Because of the implementation, the mocks need to be enhanced.

Does that help answer your question?

**Comment by bnordland** [comment]
This pull request is now up to date with the master.

File: commands/droplets.go [hunk 3]
```diff
@@ -462,33 +463,60 @@ func RunDropletDelete(c *CmdConfig) error {
...
+
 	}
-	return fmt.Errorf("operation aborted")
 
-	return nil
+	if len(c.Args) < 1 && tagName == "" {
+		return doctl.NewMissingArgsErr(c.NS)
+	} else if len(c.Args) > 0 && tagName != "" {
+		return fmt.Errorf("please specify droplets identifiers or a tag name")
+	} else if tagName != "" {
```
Comment (xmudrii) [inline commands/droplets.go]: Thank you for answer and explanation. Code is working very well. However, this is probably a point where something can be improved.
This part looks OK, but this is not needed for force action. In case of `force`, you don't want to see list and you don't need for list to be generated. In my opinion, you should check is `force` true, and if it is — exit function with `return ds.DeleteByTag(tagName)`. If it's false, execute below code as it is now.
Not something glorious, but it'll save one API call and speed up process (maybe not noticeable) because it doesn't need to query for all Droplets with tag, generate list and go through the `deleteFn` function.

About mocks, I think there is no anything to be improved. There are substituting API actions with desired response so it doesn't return error (like unauthenticated, resource doesn't exist, etc.). There are generated for code in `do` folder using [mockery](https://github.com/vektra/mockery).

**Comment by mauricio** [comment]
This, unfortunately, changes existing behavior, we could introduce a flag that will print the droplets and ask before deleting them but adding this confirmation check now will break existing callers, which is not ideal so we can't really go forward with this.
